### PR TITLE
Fix flex_attention score_mod with no score gradient

### LINKED ISSUE #145050: Make flex_attention work if `score_mod`'s output doesn't require gradients at all
### 🚀 The feature, motivation and pitch

See https://github.com/pytorch/pytorch/issues/139548#issuecomment-2597509430

```
import warnings
import numpy as np
import torch
from torch.nn.attention.flex_attention import flex_attention, create_mask, create_block_mask

# import astropy_healpix as hp


hlc = 4
num_healpix_cells = 12 * 4**hlc 
print( f'seq_length : {num_healpix_cells}')

# with warnings.catch_warnings(action="ignore"):
#     nbours= hp.neighbours( np.arange(num_healpix_cells), 2**hlc, order='nested').transpose()
# build adjacency matrix (smarter ways to do it ...)
nbours_mat = torch.zeros( (num_healpix_cells,num_healpix_cells), dtype=torch.bool, device='cuda')
# for i in range(num_healpix_cells) :
#     for j in nbours[i] :
#         nbours_mat[i,j] = True if j>=0 else False
hp_adjacency = nbours_mat

# tc_tokens = torch.from_numpy( np.load( 'tc_tokens.npy')).to(torch.float16).to('cuda')
tc_tokens = torch.ones( [204458, 256], dtype=torch.float16, device='cuda', requires_grad=True)
# tcs_lens = torch.from_numpy( np.load( './tcs_lens.npy')).to(torch.int32).to('cuda')
# tcs_lens = torch.ra
# print( f'tc_tokens = {tc_tokens.shape}')
# print( f'tcs_lens = {tcs_lens.shape}')

tc_tokens_cell_idx = torch.zeros(204458, dtype=torch.int, device='cuda')

def sparsity_mask( score, b, h, q_idx, kv_idx):
    return hp_adjacency[ tc_tokens_cell_idx[q_idx], tc_tokens_cell_idx[kv_idx] ]

compiled_flex_attention = torch.compile(flex_attention, dynamic=False)

toks = tc_tokens[:,:64].unsqueeze(0).unsqueeze(0)
out = compiled_flex_attention( toks, toks, toks, score_mod=sparsity_mask)

t = torch.zeros_like( out)
mse = torch.nn.MSELoss()
loss = mse( t, out)
loss.backward()
```

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185991

FlexAttention builds a joint graph for score_mod so the backward template can compute the gradient of the modified scores with respect to the raw attention scores. When score_mod returns a value that is independent of score, AOTAutograd correctly reports no gradient for the score input. The FlexAttention lowering did not handle that None result: Inductor still expects a score-gradient subgraph output for the dq/dk matmuls, so backward failed during lowering/kernel generation instead of treating the score gradient as zero.

Materialize a zero score gradient when the joint graph returns None for the score input. Constant joint graphs can lower that zero as a scalar/rank-1 Triton value, so the backward template now broadcasts only those low-rank grad_scores values to the score tile before the matmuls. Rank-2 gradients from normal differentiable score_mod paths skip the extra broadcast add.

I considered only adding a template-side fallback, but that would leave the higher-order op contract ambiguous: the joint graph should always provide a score-gradient value to the backward lowering. Materializing the zero in create_fw_bw_graph fixes that contract, while the template rank guard handles the generated representation needed by Triton.

Fixes #145050

Generated by my agent

Benchmark Results:

- CUDA fp16 compiled flex_attention backward, B=1 H=1 S=512 D=64, score_mod returns score * 1.1, 5 warmup iterations and 50 CUDA-event timed iterations.
- Baseline main median: 0.5343 ms; p10/p90: 0.4784/0.6628 ms.
- Patched median: 0.5404 ms; p10/p90: 0.5289/0.5629 ms.
- Result: no clear regression beyond run-to-run noise; baseline timings were bimodal in this environment, while patched timings stayed in the same envelope.

Test Plan:

- Reproduced the issue before the fix with a minimal CUDA repro using torch.compile(flex_attention, dynamic=False), score_mod returning q_idx >= kv_idx, and backward failing with InductorError / joint_subgraph_buffer is None.
- Verified the repro after the fix: q.grad and k.grad are zero, v.grad is nonzero.
- Verified backend="aot_eager" repro after the fix.
- Verified differentiable score_mod=score * score smoke test after the fix.
- python test/inductor/test_flex_attention.py -k test_score_mod_without_score_gradient
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by drisspg — CHANGES_REQUESTED** [review]
no we should error

**Review by jansel — COMMENTED** [review]
My agent says I updated the PR to error explicitly when flex_attention backward sees a score_mod output that does not depend on score, instead of materializing a zero score gradient. The new test asserts that RuntimeError path.

**Comment by jansel** [comment]
@claude review these changes

**Review by drisspg — APPROVED** [review]

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?
